File: tests/test_first_run_gate_js.py
"""The first-run report must not call a working install empty (#5766).

Reported from a hosted node page whose header read "Claude Code · 1281
sessions" while the panel underneath it said "No agent sessions on this
machine yet ... No supported runtime was detected".

Two independent faults, one guarded here per fault:

  * the count gate treated a MISSING key as a zero. `/api/overview` has no
    single canonical session-count field — OSS serves `sessions` and
    `sessionCount`, and the cloud node page builds the payload client-side
    out of the encrypted snapshot with `sessionCount` only. The behaviour is
    pinned in ``test_first_run_gate_js.js`` against the shipped source;
  * the panel is a LOCAL-machine diagnostic. It probes this machine's
    runtime paths and prescribes ``clawmetry connect`` / ``clawmetry
    --sample``. On a hosted node page the probe runs inside the cloud
    container, so it answered for the server while the reader was looking at
    their laptop. It must not render there at all.

Recorded as ADR-006 on the Sample Mode and First-Run Report blueprint
(89dfab4d-d6c1-4c72-9222-0e68acd0eeaa), which also carries the amended
scope contract: local dashboards only, and only on a payload that
positively reports zero.

The Node suite is a required gate here rather than a sibling of
``test_appjs_units.py``: that file is green locally and named in no CI job,
which is how a guard rots without anyone noticing.
"""

from __future__ import annotations

import os
import re
import shutil
import subprocess

import pytest

_HERE = os.path.dirname(os.path.abspath(__file__))
_JS_TEST = os.path.join(_HERE, "test_first_run_gate_js.js")
_APP_JS = os.path.join(_HERE, "..", "clawmetry", "static", "js", "app.js")


@pytest.mark.skipif(
    shutil.which("node") is None,
    reason="node not on PATH; JS unit tests only run when Node is available",
)
def test_first_run_gate_unit_suite() -> None:
    proc = subprocess.run(
        ["node", _JS_TEST], capture_output=True, text=True, timeout=30
    )
    output = (proc.stdout or "") + (proc.stderr or "")
    assert proc.returncode == 0, "first-run gate tests failed:\n" + output
    assert "PASS" in output, "no PASS line in output:\n" + output


def _render_body() -> str:
    src = open(_APP_JS, encoding="utf-8").read()
    m = re.search(r"^async function renderFirstRunReport\b[\s\S]*?^\}", src, re.M)
    assert m, "renderFirstRunReport not found in app.js"
    return m.group(0)


def test_panel_never_renders_on_a_hosted_node_page() -> None:
    """A local-machine diagnostic has no honest answer on cloud."""
    body = _render_body()
    gate = re.search(
        r"if \(window\.CLOUD_MODE\) \{[^}]*el\.style\.display = 'none';[^}]*return;",
        body,
    )
    assert gate, "renderFirstRunReport does not bail out under window.CLOUD_MODE"
    # And it must bail BEFORE it probes for runtimes — the probe is the part
    # that answers about the wrong machine.
    assert body.index("CLOUD_MODE") < body.index("runtime-detection"), (
        "the CLOUD_MODE bail-out must come before the runtime-detection probe"
    )


def test_the_gate_asks_looks_empty_not_a_raw_key_lookup() -> None:
    """Pins the seam so a later edit cannot re-inline a three-key guess."""
    body = _render_body()
    assert "_frrLooksEmpty(overview)" in body, (
        "the emptiness decision must go through _frrLooksEmpty"
    )


def test_session_key_list_covers_the_cloud_payload() -> None:
    """`sessionCount` is the ONLY count the cloud node page ships."""
    src = open(_APP_JS, encoding="utf-8").read()
    m = re.search(r"var _FRR_SESSION_KEYS = \[([\s\S]*?)\];", src)
    assert m, "_FRR_SESSION_KEYS not found in app.js"
    keys = set(re.findall(r"'([^']+)'", m.group(1)))
    for required in ("sessions", "sessionCount"):
        assert required in keys, f"{required} missing from _FRR_SESSION_KEYS: {keys}"

File: tests/test_first_run_gate_js.js
// The gate on the first-run report panel (#5766).
//
// The panel says "No agent sessions on this machine yet" and then tells the
// reader nothing was detected and to try `clawmetry --sample`. Shown to
// somebody whose sessions ARE syncing, every one of those sentences is a
// lie, and it renders directly under a header counting their sessions.
//
// It shipped that way because the gate read a session count out of
// `/api/overview` by trying three key names and returning 0 when none of
// them was present. The cloud node page builds that payload client-side
// from the encrypted snapshot and carries `sessionCount` only — none of the
// three — so a machine with 1,281 synced sessions scored zero and was told
// it had no agents.
//
// Runs under `node tests/test_first_run_gate_js.js` — no jsdom, ~50ms.
// Pulls the real functions out of app.js so this tests shipped source.

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const APP_JS = path.join(__dirname, '..', 'clawmetry', 'static', 'js', 'app.js');
const src = fs.readFileSync(APP_JS, 'utf8');

let passed = 0;
let failed = 0;

function eq(actual, expected, label) {
  if (actual === expected) {
    passed++;
    console.log('  ok   ' + label);
  } else {
    failed++;
    console.log('  FAIL ' + label);
    console.log('       expected: ' + JSON.stringify(expected));
    console.log('       actual:   ' + JSON.stringify(actual));
  }
}

function extractFunction(name) {
  const re = new RegExp('^function ' + name + '\\b[\\s\\S]*?^\\}', 'm');
  const m = src.match(re);
  if (!m) throw new Error('could not find function ' + name + ' in app.js');
  return m[0];
}

const keyVars = src.match(/^var _FRR_SESSION_KEYS[\s\S]*?^var _FRR_EVENT_KEYS[\s\S]*?;$/m);
if (!keyVars) throw new Error('could not find the _FRR_* key lists in app.js');

const sandbox = { isFinite: isFinite, Array: Array };
vm.createContext(sandbox);
vm.runInContext(
  keyVars[0] + '\n'
  + extractFunction('_frrCount') + '\n'
  + extractFunction('_frrLooksEmpty') + '\n'
  + 'this._empty = _frrLooksEmpty; this._count = _frrCount;'
  + 'this._SK = _FRR_SESSION_KEYS;', sandbox);
const empty = sandbox._empty;

console.log('_frrLooksEmpty — the panel only fires on a payload that says zero');

// The regression. This IS the shape the cloud node page hands the frontend.
eq(empty({ sessionCount: 1281, mainTokens: 900000, spending: { today: 4 } }), false,
   'a cloud payload carrying only sessionCount is NOT empty (#5766)');
eq(sandbox._count({ sessionCount: 1281 }, sandbox._SK), 1281,
   'sessionCount is read as a session count');

// OSS shapes, which carry `sessions` as well.
eq(empty({ sessions: 12, sessionCount: 12, sessionsToday: 3 }), false,
   'an OSS payload with sessions is not empty');
eq(empty({ sessions: [{ id: 'a' }] }), false,
   'a non-empty sessions ARRAY is not empty');

// A busy machine that has simply not run anything since midnight. Taking the
// first key found rather than the max would call this one empty.
eq(empty({ sessionsToday: 0, sessionCount: 1281 }), false,
   'sessionsToday=0 does not override a non-zero all-time count');

// Unknown is not zero. A reshaped, truncated or timed-out payload must stay
// silent rather than accuse a working install of having no runtimes.
eq(empty({}), false, 'an empty object is unknown, not empty');
eq(empty(null), false, 'a null payload is unknown, not empty');
eq(empty(undefined), false, 'an undefined payload is unknown, not empty');
eq(empty({ model: 'claude-opus-5' }), false,
   'a payload with no count field at all is unknown, not empty');

// And the panel must still do its job on a genuinely fresh machine.
eq(empty({ sessions: [], sessionCount: 0, events: 0 }), true,
   'a genuinely empty payload still shows the panel');
eq(empty({ sessions: 0, sessionCount: 0, sessionsToday: 0, events: 0 }), true,
   'all-zero counts still show the panel');
eq(empty({ sessionCount: 0 }), true,
   'a cloud payload reporting zero sessions still shows the panel');

console.log('\n' + (failed === 0 ? 'PASS' : 'FAIL') + ' — ' + passed + ' passed, ' + failed + ' failed');
process.exit(failed === 0 ? 0 : 1);
